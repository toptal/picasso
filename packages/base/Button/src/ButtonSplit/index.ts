import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './ButtonSplit'

export { default as ButtonSplit } from './ButtonSplit'
export type ButtonSplitProps = OmitInternalProps<Props>
