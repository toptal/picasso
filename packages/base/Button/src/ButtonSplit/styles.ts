import type { Theme } from '@material-ui/core/styles'
import { createStyles } from '@material-ui/core/styles'

export default ({ palette }: Theme) =>
  createStyles({
    actionButton: {
      borderTopRightRadius: 0,
      borderBottomRightRadius: 0,
    },
    menuButton: {
      minWidth: '2em',
      paddingLeft: '0.5em',
      paddingRight: '0.5em',
      borderTopLeftRadius: 0,
      borderBottomLeftRadius: 0,
      marginLeft: 0,
    },

    primaryVariant: {
      '&$actionButton': {
        borderRight: `1px solid ${palette.blue.darker}`,
      },
      '&$menuButton': {
        borderLeft: `1px solid ${palette.blue.darker}`,

        '&$disabled': {
          borderLeftColor: palette.grey.main,
        },
      },
    },

    smallSize: {
      minWidth: '1.5em',
      paddingLeft: '0.25em',
      paddingRight: '0.25em',
    },
    mediumSize: {},
    largeSize: {
      minWidth: '3em',
      paddingLeft: '0.75em',
      paddingRight: '0.75em',
    },

    dropdown: {
      display: 'block',
      cursor: 'pointer',
    },

    disabled: {
      // override dropdown anchor styles
      '& > div': {
        cursor: 'auto',
      },
    },

    rotated: {
      transform: 'rotate(180deg)',
    },
  })
