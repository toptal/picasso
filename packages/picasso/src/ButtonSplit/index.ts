import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './ButtonSplit'

export { default } from './ButtonSplit'
export type ButtonSplitProps = OmitInternalProps<Props>
