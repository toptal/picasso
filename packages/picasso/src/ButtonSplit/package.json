{
  "name": "@toptal/picasso-button-split",
  "version": "1.0.0",
  "main": "index.js",
  "dependencies": {
    "@toptal/picasso-icon": "^1.0.0",
    "@toptal/picasso-button": "^1.0.0",
    "@toptal/picasso-button-group": "^1.0.0",
    "@toptal/picasso-dropdown": "^1.0.0"
  }
}