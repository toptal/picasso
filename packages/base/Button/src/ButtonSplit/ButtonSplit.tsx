import type { ReactNode, MouseEvent, HTMLAttributes } from 'react'
import React, { forwardRef } from 'react'
import type { SizeType, BaseProps } from '@toptal/picasso-shared'
import { twMerge } from '@toptal/picasso-tailwind-merge'
import { ArrowDownMinor24, ArrowDownMinor16 } from '@toptal/picasso-icons'
import { Dropdown } from '@toptal/picasso-dropdown'

import type { ButtonProps } from '../Button'
import { Button } from '../Button'
import { ButtonGroup } from '../ButtonGroup'
import { createButtonGroupItemClassNames } from '../ButtonGroupItem/styles'
import {
  createActionButtonClassNames,
  createMenuButtonClassNames,
} from './styles'

export interface Props
  extends BaseProps,
    Omit<HTMLAttributes<HTMLDivElement>, 'onClick'> {
  /** Content of Button component */
  children: ReactNode
  /** Content element that opens when anchor is clicked */
  menu: ReactNode
  /** Callback invoked when component is clicked */
  onClick?: ButtonProps['onClick']
  /** A button can have different sizes */
  size?: ButtonProps['size']
  /** The variant to use */
  variant?: 'primary' | 'secondary'
  /** Is component disaled or not */
  disabled?: boolean
  // Internal props
  menuButtonProps?: Omit<ButtonProps, 'children'>
  actionButtonProps?: Omit<ButtonProps, 'children'>
  testIds?: {
    actionButton?: string
    menuButton?: string
  }
}

const DropdownIcon = ({
  size,
  className,
}: {
  size: SizeType<'small' | 'medium' | 'large'>
  className?: string
}) => {
  if (size === 'large') {
    return <ArrowDownMinor24 className={className} />
  }

  return <ArrowDownMinor16 className={className} />
}

const EventStopPropagation = ({ children }: { children: ReactNode }) => {
  const handleClick = (
    event: MouseEvent<HTMLButtonElement & HTMLAnchorElement>
  ) => {
    event.stopPropagation()
  }

  return <span onClick={handleClick}>{children}</span>
}

export const ButtonSplit = forwardRef<HTMLDivElement, Props>(
  function ButtonSplit(props, ref) {
    const {
      size = 'medium',
      menu,
      children,
      variant = 'primary',
      disabled,
      style,
      className,
      onClick,
      menuButtonProps,
      actionButtonProps,
      testIds = {},
      ...rest
    } = props

    const renderMenuButton = ({ isOpen }: { isOpen: boolean }) => {
      const menuButtonClassName = twMerge(
        createButtonGroupItemClassNames({
          active: menuButtonProps?.active,
          hovered: menuButtonProps?.hovered,
          disabled: menuButtonProps?.disabled || disabled,
          focused: menuButtonProps?.focused,
        }),
        createMenuButtonClassNames({
          variant,
          size,
          disabled: menuButtonProps?.disabled || disabled,
        }),
        menuButtonProps?.className
      )

      const iconClassName = isOpen ? 'rotate-180' : ''

      const menuButton = (
        <Button
          {...menuButtonProps}
          variant={variant}
          className={menuButtonClassName}
          size={size}
          disabled={disabled}
          data-testid={testIds.menuButton}
        >
          <DropdownIcon className={iconClassName} size={size} />
        </Button>
      )

      return disabled ? (
        <EventStopPropagation>{menuButton}</EventStopPropagation>
      ) : (
        menuButton
      )
    }

    const actionButtonClassName = twMerge(
      createButtonGroupItemClassNames({
        active: actionButtonProps?.active,
        hovered: actionButtonProps?.hovered,
        disabled: actionButtonProps?.disabled || disabled,
        focused: actionButtonProps?.focused,
      }),
      createActionButtonClassNames({
        variant,
      }),
      actionButtonProps?.className
    )

    const dropdownClassName = twMerge(
      'block cursor-pointer',
      disabled && '[&>div]:cursor-auto'
    )

    return (
      <ButtonGroup {...rest} ref={ref} style={style} className={className}>
        <Button
          {...actionButtonProps}
          className={actionButtonClassName}
          size={size}
          variant={variant}
          disabled={disabled}
          onClick={onClick}
          data-testid={testIds.actionButton}
        >
          {children}
        </Button>
        <Dropdown content={menu} className={dropdownClassName}>
          {({ isOpen }: { isOpen: boolean }) => renderMenuButton({ isOpen })}
        </Dropdown>
      </ButtonGroup>
    )
  }
)

ButtonSplit.defaultProps = {
  size: 'medium',
  variant: 'primary',
}

ButtonSplit.displayName = 'ButtonSplit'

export default ButtonSplit
