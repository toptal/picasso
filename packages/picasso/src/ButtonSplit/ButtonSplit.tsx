import React, { forwardRef, ReactNode, MouseEvent, HTMLAttributes } from 'react'
import { makeStyles, Theme } from '@material-ui/core/styles'
import { SizeType, BaseProps } from '@toptal/picasso-shared'
import cx from 'classnames'

import { ArrowDownMinor24, ArrowDownMinor16 } from '../Icon'
import Button, { ButtonProps } from '../Button'
import Dropdown from '../Dropdown'
import styles from './styles'

export interface Props
  extends BaseProps,
    Omit<HTMLAttributes<HTMLDivElement>, 'onClick'> {
  /** Content of Button component */
  children: ReactNode
  /** Content element that opens when anchor is clicked */
  menu: ReactNode
  /** Callback invoked when component is clicked */
  onClick?: ButtonProps['onClick']
  /** A button can have different sizes */
  size?: ButtonProps['size']
  /** The variant to use */
  variant?: 'primary' | 'secondary'
  /** Is component disaled or not */
  disabled?: boolean
  // Internal props
  menuButtonProps?: Omit<ButtonProps, 'children'>
  actionButtonProps?: Omit<ButtonProps, 'children'>
  testIds?: {
    actionButton?: string
    menuButton?: string
  }
}
// Using { index: -1 } to inject CSS link to the bottom of the head
// in order to prevent Button's styles to override ButtonSplit's ones
// Related Jira issue: https://toptal-core.atlassian.net/browse/FX-1520
const useStyles = makeStyles<Theme>(styles, {
  name: 'PicassoButtonSplit',
  index: -1
})

const DropdownIcon = ({
  size,
  className
}: {
  size: SizeType<'small' | 'medium' | 'large'>
  className?: string
}) => {
  if (size === 'large') {
    return <ArrowDownMinor24 className={className} />
  }

  return <ArrowDownMinor16 className={className} />
}

const EventStopPropagation = ({ children }: { children: ReactNode }) => {
  const handleClick = (
    event: MouseEvent<HTMLButtonElement & HTMLAnchorElement>
  ) => {
    event.stopPropagation()
  }

  return <span onClick={handleClick}>{children}</span>
}

export const ButtonSplit = forwardRef<HTMLDivElement, Props>(
  function ButtonSplit(props, ref) {
    const {
      size = 'medium',
      menu,
      children,
      variant = 'primary',
      disabled,
      style,
      className,
      onClick,
      menuButtonProps,
      actionButtonProps,
      testIds = {},
      ...rest
    } = props
    const classes = useStyles()

    const commonClasses = cx(classes.button, {
      [classes.primaryVariant]: variant === 'primary',
      [classes.disabled]: disabled
    })

    const renderMenuButton = ({
      isOpen,
      disabled
    }: {
      isOpen: boolean
      disabled?: boolean
    }) => {
      const menuButton = (
        <Button
          {...menuButtonProps}
          variant={variant}
          className={`${commonClasses} ${classes.menuButton} ${
            classes[`${size}Size`]
          }`}
          size={size}
          disabled={disabled}
          data-testid={testIds.menuButton}
        >
          <DropdownIcon
            className={cx({
              [classes.rotated]: isOpen
            })}
            size={size}
          />
        </Button>
      )

      return disabled ? (
        <EventStopPropagation>{menuButton}</EventStopPropagation>
      ) : (
        menuButton
      )
    }

    return (
      <Button.Group {...rest} ref={ref} style={style} className={className}>
        <Button
          {...actionButtonProps}
          className={`${commonClasses} ${classes.actionButton}`}
          size={size}
          variant={variant}
          disabled={disabled}
          onClick={onClick}
          data-testid={testIds.actionButton}
        >
          {children}
        </Button>
        <Dropdown
          content={menu}
          className={cx(classes.dropdown, {
            [classes.disabled]: disabled
          })}
        >
          {({ isOpen }: { isOpen: boolean }) =>
            renderMenuButton({ isOpen, disabled })
          }
        </Dropdown>
      </Button.Group>
    )
  }
)

ButtonSplit.defaultProps = {
  size: 'medium',
  variant: 'primary'
}

ButtonSplit.displayName = 'ButtonSplit'

export default ButtonSplit
