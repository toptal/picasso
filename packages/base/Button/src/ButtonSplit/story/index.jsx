import { ButtonSplit } from '../ButtonSplit'
import PicassoBook from '~/.storybook/components/PicassoBook'

const chapter = PicassoBook.connectToPage(page =>
  page
    .createChapter('Split Button', 'Combine dropdown with menu and a button.')
    .addExample(
      'ButtonSplit/story/Variants.example.tsx',
      'Variants',
      'base/Button'
    )
    .addExample('ButtonSplit/story/Sizes.example.tsx', 'Sizes', 'base/Button')
    // happo makes snapshot of the DOM and replicates it on their machine to make a screenshot
    // so focused, active and other states of the button are lost
    .addExample('ButtonSplit/story/States.example.tsx', 'States', 'base/Button')
)

const componentDocs = PicassoBook.createComponentDocs(
  ButtonSplit,
  'Button.Split'
)

export default {
  chapter,
  componentDocs,
}
