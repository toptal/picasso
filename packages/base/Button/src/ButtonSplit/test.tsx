import React from 'react'
import { render, fireEvent, screen } from '@toptal/picasso-test-utils'
import { MenuCompound as PicassoMenu } from '@toptal/picasso-menu'

import { ButtonSplit } from './ButtonSplit'

const Menu = () => {
  const handleClick = () => {}

  return (
    <PicassoMenu data-testid='menu'>
      <PicassoMenu.Item onClick={handleClick}>First item</PicassoMenu.Item>
      <PicassoMenu.Item onClick={handleClick}>Second item</PicassoMenu.Item>
      <PicassoMenu.Item onClick={handleClick}>Third item</PicassoMenu.Item>
    </PicassoMenu>
  )
}

const smallIconViewBox = '0 0 16 16'
const largeIconViewBox = '0 0 24 24'

describe('ButtonSplit', () => {
  it('default render', () => {
    const { container } = render(
      <ButtonSplit menu={<Menu />}>Action</ButtonSplit>
    )

    expect(container).toMatchSnapshot()
  })

  it('invokes onClick after a click on action button', () => {
    const onClick = jest.fn()

    render(
      <ButtonSplit
        onClick={onClick}
        testIds={{ actionButton: 'action' }}
        menu={<Menu />}
      >
        Action
      </ButtonSplit>
    )

    const button = screen.getByTestId('action')

    fireEvent.click(button)

    expect(onClick).toHaveBeenCalled()
  })

  it('renders arrow up icon for an opened menu', () => {
    render(
      <ButtonSplit testIds={{ menuButton: 'menu' }} menu={<Menu />}>
        Action
      </ButtonSplit>
    )

    const button = screen.getByTestId('menu')

    // open menu
    fireEvent.click(button)

    expect(button.querySelector('svg')).toMatchSnapshot()
  })

  it('renders arrow down icon for a closed menu', () => {
    render(
      <ButtonSplit testIds={{ menuButton: 'menu' }} menu={<Menu />}>
        Action
      </ButtonSplit>
    )

    const button = screen.getByTestId('menu')

    // open menu
    fireEvent.click(button)

    // close menu
    fireEvent.click(button)

    expect(button.querySelector('svg')).toMatchSnapshot()
  })

  it('renders small arrow for small button', () => {
    render(
      <ButtonSplit
        testIds={{ menuButton: 'menu' }}
        size='small'
        menu={<Menu />}
      >
        Action
      </ButtonSplit>
    )

    expect(screen.getByTestId('menu').querySelector('svg')).toHaveAttribute(
      'viewBox',
      smallIconViewBox
    )
  })

  it('renders small arrow for medium button', () => {
    render(
      <ButtonSplit
        testIds={{ menuButton: 'menu' }}
        size='medium'
        menu={<Menu />}
      >
        Action
      </ButtonSplit>
    )

    expect(screen.getByTestId('menu').querySelector('svg')).toHaveAttribute(
      'viewBox',
      smallIconViewBox
    )
  })

  it('renders small arrow for large button', () => {
    render(
      <ButtonSplit
        testIds={{ menuButton: 'menu' }}
        size='large'
        menu={<Menu />}
      >
        Action
      </ButtonSplit>
    )

    expect(screen.getByTestId('menu').querySelector('svg')).toHaveAttribute(
      'viewBox',
      largeIconViewBox
    )
  })

  it('closes menu on click to the item', async () => {
    render(
      <ButtonSplit
        testIds={{ menuButton: 'menu-button', actionButton: 'action' }}
        size='large'
        menu={<Menu />}
      >
        Action
      </ButtonSplit>
    )

    const menuButton = screen.getByTestId('menu-button')

    // open menu
    fireEvent.click(menuButton)

    // click away
    fireEvent.click(
      screen.getByRole('menuitem', {
        name: /first item/i,
      })
    )

    expect(
      screen.getByTestId('menu-button').querySelector('svg')
    ).toMatchSnapshot()
  })

  it('invokes a menu item handler on click to the item', () => {
    const handleClick = jest.fn()

    const menu = (
      <PicassoMenu data-testid='menu'>
        <PicassoMenu.Item onClick={handleClick}>First item</PicassoMenu.Item>
        <PicassoMenu.Item onClick={handleClick}>Second item</PicassoMenu.Item>
        <PicassoMenu.Item onClick={handleClick}>Third item</PicassoMenu.Item>
      </PicassoMenu>
    )

    render(
      <ButtonSplit
        testIds={{ menuButton: 'menu-button', actionButton: 'action' }}
        size='large'
        menu={menu}
      >
        Action
      </ButtonSplit>
    )

    const menuButton = screen.getByTestId('menu-button')

    // open menu
    fireEvent.click(menuButton)

    // click away
    fireEvent.click(
      screen.getByRole('menuitem', {
        name: /first item/i,
      })
    )

    expect(handleClick).toHaveBeenCalledTimes(1)
  })
})
