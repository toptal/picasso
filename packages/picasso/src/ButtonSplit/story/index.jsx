import { ButtonSplit } from '../ButtonSplit'
import PicassoBook from '~/.storybook/components/PicassoBook'

const chapter = PicassoBook.connectToPage(page =>
  page
    .createChapter('Split Button', 'Combine dropdown with menu and a button.')
    .addExample('ButtonSplit/story/Variants.example.tsx', 'Variants')
    .addExample('ButtonSplit/story/Sizes.example.tsx', 'Sizes')
    // happo makes snapshot of the DOM and replicates it on their machine to make a screenshot
    // so focused, active and other states of the button are lost
    .addExample('ButtonSplit/story/States.example.tsx', 'States')
)

const componentDocs = PicassoBook.createComponentDocs(
  ButtonSplit,
  'Button.Split'
)

export default {
  chapter,
  componentDocs,
}
