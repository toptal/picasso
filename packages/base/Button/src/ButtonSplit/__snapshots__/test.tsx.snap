// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonSplit closes menu on click to the item 1`] = `
<svg
  class="PicassoSvgArrowDownMinor24-root-24"
  style="min-width: 24px; min-height: 24px;"
  viewBox="0 0 24 24"
>
  <path
    d="m18.007 8.3.707.707-6 6-.707.707-.707-.707-6-6 .707-.707 6 6 6-6Z"
  />
</svg>
`;

exports[`ButtonSplit default render 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="flex justify-start [&>[data-component-type="button"]:hover]:z [&>[data-component-type="button"]:first-child:not(:last-child)]:rounded-r [&>[data-component-type="button"]:first-child:not(:last-child)]:ml [&>[data-component-type="button"]:not(:first-child):not(:last-child)]:rounded [&>[data-component-type="button"]:not(:first-child):not(:last-child)]:ml-[ [&>[data-component-type="button"]:last-child:not(:first-child)]:rounded-l [&>[data-component-type="button"]:last-child:not(:first-child)]:ml-[ [&_:first-child:not(:last-child)_[data-component-type="button"]]:rounded-r [&_:first-child:not(:last-child)_[data-component-type="button"]]:ml [&_:not(:first-child):not(:last-child)_[data-component-type="button"]]:rounded [&_:not(:first-child):not(:last-child)_[data-component-type="button"]]:ml-[ [&_:last-child:not(:first-child)_[data-component-type="button"]]:rounded-l [&_:last-child:not(:first-child)_[data-component-type="button"]]:ml-["
    >
      <button
        aria-disabled="false"
        class="base-Button text-lg inline-flex items-center justify-center select-none appearance-none m-0 relative normal-case align-middle duration-350 ease-out shrink-0 outline-none [&+&]:ml-4 cursor-pointer no-underline hover:no-underline rounded-sm shadow-none focus-visible:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] focus-within:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] text-white visited:text-white hover:bg-[#4269D6] active:bg-[#1A41AB] bg-blue min-w h-8 py-0 px-4 active:z-[1] hover:z-[1] focus-visible:z-[1] disabled:z-[1] transition-[color,background] border-r border-l border-y border-solid border-blue"
        data-component-type="button"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="items-center inline-flex font-semibold whitespace-nowrap text-button"
        >
          Action
        </span>
      </button>
      <div
        class="items-center block cursor-pointer"
      >
        <div
          class="inline-flex items-center cursor-pointer"
        >
          <button
            aria-disabled="false"
            class="base-Button text-lg inline-flex items-center justify-center select-none appearance-none m-0 relative normal-case align-middle duration-350 ease-out shrink-0 outline-none [&+&]:ml-4 cursor-pointer no-underline hover:no-underline rounded-sm shadow-none focus-visible:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] focus-within:shadow-[0_0_0_3px_rgba(32,78,207,0.48)] text-white visited:text-white hover:bg-[#4269D6] active:bg-[#1A41AB] bg-blue h-8 py-0 transition-[color,background] active:z-[1] hover:z-[1] focus-visible:z-[1] disabled:z-[1] min-w px-[0.5em] border-l border-r border-y border-solid border-blue"
            data-component-type="button"
            role="button"
            tabindex="0"
            type="button"
          >
            <span
              class="items-center inline-flex font-semibold whitespace-nowrap text-button"
            >
              <svg
                class="PicassoSvgArrowDownMinor16-root"
                style="min-width: 16px; min-height: 16px;"
                viewBox="0 0 16 16"
              >
                <path
                  d="m11.997 5.29.707.707-4 4-.707.707-.707-.707-4-4 .707-.707 4 4 4-4Z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ButtonSplit renders arrow down icon for a closed menu 1`] = `
<svg
  class="PicassoSvgArrowDownMinor16-root-24"
  style="min-width: 16px; min-height: 16px;"
  viewBox="0 0 16 16"
>
  <path
    d="m11.997 5.29.707.707-4 4-.707.707-.707-.707-4-4 .707-.707 4 4 4-4Z"
  />
</svg>
`;

exports[`ButtonSplit renders arrow up icon for an opened menu 1`] = `
<svg
  class="PicassoSvgArrowDownMinor16-root-24 rotate-180"
  style="min-width: 16px; min-height: 16px;"
  viewBox="0 0 16 16"
>
  <path
    d="m11.997 5.29.707.707-4 4-.707.707-.707-.707-4-4 .707-.707 4 4 4-4Z"
  />
</svg>
`;
